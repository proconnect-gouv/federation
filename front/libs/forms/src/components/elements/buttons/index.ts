export * from './array-add';
export * from './array-remove';
export * from './clipboard';
