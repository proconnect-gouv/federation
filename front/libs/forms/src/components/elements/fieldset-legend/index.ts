/* istanbul ignore file */

// declarative file
export * from './fieldset-legend.element';
