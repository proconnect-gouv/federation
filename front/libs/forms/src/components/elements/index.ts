export * from './fieldset';
export * from './fieldset-legend';
export * from './group';
export * from './label';
export * from './message';
