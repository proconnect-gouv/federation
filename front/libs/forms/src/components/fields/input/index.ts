/* istanbul ignore file */

// declarative file
export * from './input.field';
