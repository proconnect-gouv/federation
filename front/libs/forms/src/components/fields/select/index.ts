/* istanbul ignore file */

// declarative file
export * from './select.field';
