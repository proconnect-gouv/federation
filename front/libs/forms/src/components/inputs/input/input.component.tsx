import React from 'react';
import type { FieldInputProps } from 'react-final-form';

import type { PropsWithClassName } from '@fc/common';

interface InputComponentProps<FieldValue = string> extends PropsWithClassName {
  readOnly?: boolean;
  disabled?: boolean;
  id: string;
  input: FieldInputProps<FieldValue, HTMLElement | HTMLSelectElement>;
}

export const InputComponent = React.memo(
  ({ className, disabled = false, id, input, readOnly = false }: InputComponentProps) => (
    <input
      {...input}
      aria-describedby={`${input.name}-messages`}
      className={className}
      data-testid={`${id}--testid`}
      disabled={disabled}
      id={id}
      name={input.name}
      readOnly={readOnly}
      type={input.type}
    />
  ),
);

InputComponent.displayName = 'InputComponent';
