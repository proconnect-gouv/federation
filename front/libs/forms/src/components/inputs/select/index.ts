/* istanbul ignore file */

// declarative file
export * from './select.input';
