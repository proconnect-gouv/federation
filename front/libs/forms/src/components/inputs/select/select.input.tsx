import React, { useMemo } from 'react';

import { t } from '@fc/i18n';

import { ComponentTypes } from '../../../enums';
import { useFieldMeta } from '../../../hooks';
import type { PropsWithInputChoicesType } from '../../../types';
import { GroupElement, LabelElement, MessageElement } from '../../elements';

export const SelectInput = React.memo(
  ({ choices, config, input, meta }: PropsWithInputChoicesType) => {
    const { hint, label, required } = config;
    const { className, disabled, name } = input;

    const { errorMessage, hasError, isValid } = useFieldMeta(meta);

    const childs = useMemo(
      () =>
        choices.map((choice) => (
          <option key={`select::${name}::option::${choice.value}`} value={choice.value}>
            {choice.label}
          </option>
        )),
      [choices, name],
    );

    return (
      <GroupElement
        className={className}
        disabled={disabled}
        hasError={hasError}
        isValid={isValid}
        type={ComponentTypes.SELECT}>
        <LabelElement hint={hint} label={label} name={name} required={required} />
        <select {...input} className="fr-select fr-select--valid">
          <option disabled hidden value="">
            {t('Form.select.placeholder')}
          </option>
          {childs}
        </select>
        <MessageElement error={errorMessage} id={name} isValid={isValid} />
      </GroupElement>
    );
  },
);

SelectInput.displayName = 'SelectInput';
