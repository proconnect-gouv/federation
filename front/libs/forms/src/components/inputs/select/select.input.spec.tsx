import { render } from '@testing-library/react';

import { t } from '@fc/i18n';

import { useFieldMeta } from '../../../hooks';
import { GroupElement, LabelElement, MessageElement } from '../../elements';
import { SelectInput } from './select.input';

jest.mock('../../elements/group/group.element');
jest.mock('../../elements/label/label.element');
jest.mock('../../elements/message/message.element');
jest.mock('../../../hooks/field-meta/field-meta.hook');

describe('SelectInput', () => {
  // Given
  const inputMock = {
    onBlur: jest.fn(),
    onChange: jest.fn(),
    onFocus: jest.fn(),
    value: 'any-value-mock',
  };
  const choices = [
    { label: 'mock-label-1', value: 'mock-value-1' },
    { label: 'mock-label-2', value: 'mock-value-2' },
  ];

  it('should match the snapshot when is valid', () => {
    // Given
    jest.mocked(useFieldMeta).mockReturnValueOnce({
      errorMessage: undefined,
      hasError: false,
      inputClassname: 'any-input-classname-mock',
      isValid: true,
    });

    // When
    const { container } = render(
      <SelectInput
        choices={choices}
        config={{ hint: 'any-hint-mock', label: 'any-label-mock', required: true }}
        input={{
          ...inputMock,
          className: 'any-className-mock',
          disabled: false,
          name: 'any-name-mock',
        }}
        meta={{ error: undefined, touched: false }}
      />,
    );

    // Then
    expect(container).toMatchSnapshot();
    expect(useFieldMeta).toHaveBeenCalledOnce();
    expect(useFieldMeta).toHaveBeenCalledWith({ error: undefined, touched: false });
    expect(GroupElement).toHaveBeenCalledOnce();
    expect(GroupElement).toHaveBeenCalledWith(
      {
        children: expect.any(Object),
        className: 'any-className-mock',
        disabled: false,
        hasError: false,
        isValid: true,
        type: 'select',
      },
      {},
    );
    expect(LabelElement).toHaveBeenCalledOnce();
    expect(LabelElement).toHaveBeenCalledWith(
      {
        hint: 'any-hint-mock',
        label: 'any-label-mock',
        name: 'any-name-mock',
        required: true,
      },
      {},
    );
    expect(MessageElement).toHaveBeenCalledOnce();
    expect(MessageElement).toHaveBeenCalledWith(
      {
        error: undefined,
        id: 'any-name-mock',
        isValid: true,
      },
      {},
    );
    expect(t).toHaveBeenCalledOnce();
    expect(t).toHaveBeenCalledWith('Form.select.placeholder');
  });

  it('should match the snapshot when has error', () => {
    // Given
    jest.mocked(useFieldMeta).mockReturnValueOnce({
      errorMessage: 'any-error-message-mock',
      hasError: true,
      inputClassname: 'any-input-classname-mock',
      isValid: false,
    });

    // When
    const { container } = render(
      <SelectInput
        choices={choices}
        config={{ hint: 'any-hint-mock', label: 'any-label-mock', required: true }}
        input={{
          ...inputMock,
          className: 'any-className-mock',
          disabled: false,
          name: 'any-name-mock',
        }}
        meta={{ error: 'any-error-message-mock', touched: false }}
      />,
    );

    // Then
    expect(container).toMatchSnapshot();
    expect(useFieldMeta).toHaveBeenCalledOnce();
    expect(useFieldMeta).toHaveBeenCalledWith({ error: 'any-error-message-mock', touched: false });
    expect(GroupElement).toHaveBeenCalledOnce();
    expect(GroupElement).toHaveBeenCalledWith(
      {
        children: expect.any(Object),
        className: 'any-className-mock',
        disabled: false,
        hasError: true,
        isValid: false,
        type: 'select',
      },
      {},
    );
    expect(LabelElement).toHaveBeenCalledOnce();
    expect(LabelElement).toHaveBeenCalledWith(
      {
        hint: 'any-hint-mock',
        label: 'any-label-mock',
        name: 'any-name-mock',
        required: true,
      },
      {},
    );
    expect(MessageElement).toHaveBeenCalledOnce();
    expect(MessageElement).toHaveBeenCalledWith(
      {
        error: 'any-error-message-mock',
        id: 'any-name-mock',
        isValid: false,
      },
      {},
    );
    expect(t).toHaveBeenCalledOnce();
    expect(t).toHaveBeenCalledWith('Form.select.placeholder');
  });

  it('should render 2 select.option', () => {
    // Given
    jest.mocked(useFieldMeta).mockReturnValueOnce({
      errorMessage: 'any-error-message-mock',
      hasError: true,
      inputClassname: 'any-input-classname-mock',
      isValid: false,
    });

    // When
    const { getByText } = render(
      <SelectInput
        choices={choices}
        config={{ hint: 'any-hint-mock', label: 'any-label-mock', required: true }}
        input={{
          ...inputMock,
          className: 'any-className-mock',
          disabled: false,
          name: 'any-name-mock',
        }}
        meta={{ error: 'any-error-message-mock', touched: false }}
      />,
    );
    const optionElt1 = getByText('mock-label-1');
    const optionElt2 = getByText('mock-label-2');

    // Then
    expect(optionElt1).toBeInTheDocument();
    expect(optionElt1).toHaveAttribute('value', 'mock-value-1');
    expect(optionElt2).toBeInTheDocument();
    expect(optionElt2).toHaveAttribute('value', 'mock-value-2');
  });
});
