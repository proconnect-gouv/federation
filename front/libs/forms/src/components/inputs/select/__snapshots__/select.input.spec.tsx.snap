// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput should match the snapshot when has error 1`] = `
<div>
  <div
    data-mockid="GroupElement"
  >
    <div
      data-mockid="LabelElement"
    />
    <select
      class="fr-select fr-select--valid"
      name="any-name-mock"
    >
      <option
        disabled=""
        hidden=""
        value=""
      >
        Form.select.placeholder
      </option>
      <option
        value="mock-value-1"
      >
        mock-label-1
      </option>
      <option
        value="mock-value-2"
      >
        mock-label-2
      </option>
    </select>
    <div
      data-mockid="MessageElement"
    />
  </div>
</div>
`;

exports[`SelectInput should match the snapshot when is valid 1`] = `
<div>
  <div
    data-mockid="GroupElement"
  >
    <div
      data-mockid="LabelElement"
    />
    <select
      class="fr-select fr-select--valid"
      name="any-name-mock"
    >
      <option
        disabled=""
        hidden=""
        value=""
      >
        Form.select.placeholder
      </option>
      <option
        value="mock-value-1"
      >
        mock-label-1
      </option>
      <option
        value="mock-value-2"
      >
        mock-label-2
      </option>
    </select>
    <div
      data-mockid="MessageElement"
    />
  </div>
</div>
`;
